File: frontend/src/modules/research-materials/researchAnalysisModel.ts
export type AnalysisAnnotationKind = 'descriptive' | 'researcher_reflection'
export type AnalysisRecordStatus = 'candidate' | 'confirmed' | 'rejected'
export type AnalysisCodingPlanStatus = 'candidate' | 'applied' | 'partially_applied' | 'rejected'
export type AnalysisCodingPlanItemStatus = 'candidate' | 'applied' | 'rejected'
export type AnalysisMemoKind = 'descriptive' | 'reflexive' | 'analytic' | 'methodological'
export type ComparisonFindingKind = 'support' | 'counterexample' | 'contradict' | 'competing_explanation' | 'evidence_gap'
export type CodebookLifecycle = 'active' | 'merged' | 'split' | 'retired'
export type MemoTargetKind = 'project' | 'material' | 'source' | 'code' | 'case' | 'comparison' | 'draft'
export type MatrixSubjectKind = 'code' | 'theme'
export type QualitativeMethod = 'thematic_analysis' | 'grounded_theory' | 'ethnography' | 'case_study' | 'narrative_research' | 'discourse_conversation_analysis' | 'literature_review'

export type ResearchAnalysisLocator = {
  block_index: number | null
  char_end: number | null
  char_start: number | null
  line_end: number | null
  line_start: number | null
  page: number | null
  paragraph: number | null
  section_path: string[]
}

export type AnalysisAnnotation = {
  annotation_id: string
  annotation_kind: AnalysisAnnotationKind
  case_label: string | null
  created_at: string
  locator: ResearchAnalysisLocator
  material_id: string
  note: string
  observed_at: string | null
  parse_id: string
  quote: string | null
  quote_end: number
  quote_hash: string
  quote_start: number
  reflection: string | null
  segment_content_hash: string
  segment_id: string
  source_available: boolean
  task_id: string
  unavailable_reason: string | null
}

export type AnalysisCode = {
  agent_run_id: string | null
  agent_turn_id: string | null
  annotation_ids: string[]
  code_id: string
  conversation_id: string | null
  created_at: string
  decided_at: string | null
  decision_reason: string | null
  definition: string
  label: string
  rationale: string
  source: string
  status: AnalysisRecordStatus
  task_id: string
  tool_call_id: string | null
  version: number
}

export type AnalysisCodingPlanItem = {
  item_id: string
  material_id: string
  parse_id: string
  segment_id: string
  segment_content_hash: string
  quote: string
  quote_hash: string
  quote_start: number
  quote_end: number
  locator: ResearchAnalysisLocator
  code_id: string
  code_label: string
  code_definition: string
  codebook_version: number | null
  confidence: number
  rationale: string
  status: string
  annotation_id: string | null
  decision_reason: string | null
}

export type AnalysisCodingPlan = {
  plan_id: string
  task_id: string
  title: string
  rationale: string
  items: AnalysisCodingPlanItem[]
  source: string
  status: string
  version: number
  created_at: string
  conversation_id: string | null
  agent_run_id: string | null
  agent_turn_id: string | null
  tool_call_id: string | null
  decided_at: string | null
  decision_reason: string | null
}

export type AnalysisMemo = {
  agent_run_id: string | null
  agent_turn_id: string | null
  annotation_ids: string[]
  code_ids: string[]
  content: string
  conversation_id: string | null
  created_at: string
  decided_at: string | null
  decision_reason: string | null
  memo_id: string
  memo_kind: AnalysisMemoKind
  source: string
  status: AnalysisRecordStatus
  task_id: string
  title: string
  tool_call_id: string | null
  version: number
}

export type ComparisonFinding = {
  annotation_ids?: string[]
  kind: ComparisonFindingKind
  statement: string
}

export type NextResearchStep = {
  action: string
  kind: string
  priority?: string
}

export type CaseComparison = {
  agent_run_id: string | null
  agent_turn_id: string | null
  case_labels: string[]
  comparison_id: string
  competing_explanations: string[]
  conversation_id: string | null
  created_at: string
  decided_at: string | null
  decision_reason: string | null
  evidence_gaps: string[]
  findings: ComparisonFinding[]
  next_steps: NextResearchStep[]
  question: string
  source: string
  status: AnalysisRecordStatus
  task_id: string
  theory_implication: string
  time_labels: string[]
  title: string
  tool_call_id: string | null
  version: number
}

export type ResearchAnalysisSnapshot = {
  annotations: AnalysisAnnotation[]
  codes: AnalysisCode[]
  comparisons: CaseComparison[]
  memos: AnalysisMemo[]
  task_id: string
  workspace?: QualitativeWorkspaceSnapshot
  method_presets?: QualitativeMethodPreset[]
  coding_plans?: AnalysisCodingPlan[]
}

export type CodebookEntry = {
  code_id: string
  inclusion_rules: string[]
  exclusion_rules: string[]
  parent_code_id: string | null
  positive_example_annotation_ids: string[]
  negative_example_annotation_ids: string[]
  lifecycle: CodebookLifecycle
  related_code_ids: string[]
  version: number
  updated_at: string
  revision_reason: string
}

export type AnalysisTheme = {
  theme_id: string
  label: string
  central_concept: string
  code_ids: string[]
  annotation_ids: string[]
  source: string
  status: AnalysisRecordStatus
  version: number
  created_at: string
  decided_at: string | null
  decision_reason: string | null
}

export type AnalysisMemoLink = {
  link_id: string
  memo_id: string
  target_kind: MemoTargetKind
  target_ref: string
  annotation_ids: string[]
  created_at: string
}

export type AnalysisCaseAttribute = { name: string; value: string }

export type AnalysisCaseProfile = {
  profile_id: string
  case_ref: string
  display_label: string
  attributes: AnalysisCaseAttribute[]
  summary: string
  annotation_ids: string[]
  memo_ids: string[]
  version: number
  updated_at: string
}

export type CaseThemeMatrixCell = {
  cell_id: string
  case_profile_id: string
  subject_kind: MatrixSubjectKind
  subject_id: string
  summary: string
  annotation_ids: string[]
  memo_ids: string[]
  finding_kinds: ComparisonFindingKind[]
  version: number
  updated_at: string
}

export type MethodPresetSelection = {
  method: QualitativeMethod
  version: number
  updated_at: string
}

export type QualitativeMethodPreset = {
  method: QualitativeMethod
  label: string
  primary_view: string
  matrix_axes: string[]
  prompts: string
  guardrails: string
}

export type QualitativeWorkspaceSnapshot = {
  schema_version: string
  content_hash: string
  method_preset: MethodPresetSelection
  codebook_entries: CodebookEntry[]
  memo_links: AnalysisMemoLink[]
  case_profiles: AnalysisCaseProfile[]
  formal_themes: AnalysisTheme[]
  candidate_themes: AnalysisTheme[]
  matrix_cells: CaseThemeMatrixCell[]
}

export type CreateAnalysisAnnotationInput = {
  annotation_kind: AnalysisAnnotationKind
  case_label?: string | null
  material_id: string
  note: string
  observed_at?: string | null
  parse_id: string
  quote_end: number
  quote_start: number
  reflection?: string | null
  segment_id: string
}

export type CreateAnalysisCodeInput = {
  annotation_ids: string[]
  definition: string
  label: string
  rationale: string
}

export type CreateAnalysisMemoInput = {
  annotation_ids?: string[]
  code_ids?: string[]
  content: string
  memo_kind: AnalysisMemoKind
  title: string
}

export type CreateCaseComparisonInput = {
  case_labels: string[]
  competing_explanations?: string[]
  evidence_gaps?: string[]
  findings: ComparisonFinding[]
  next_steps?: NextResearchStep[]
  question: string
  theory_implication: string
  time_labels?: string[]
  title: string
}

export type DecideAnalysisRecordInput = {
  decision: AnalysisRecordStatus
  expected_version: number
  reason: string
}

export type DecideCodingPlanInput = {
  expected_version: number
  decisions: Array<{ item_id: string; decision: 'confirmed' | 'rejected'; reason: string }>
}

export type RevokeCodingPlanInput = {
  expected_version: number
  reason: string
}

export type ConfigureCodebookEntryInput = {
  expected_version: number | null
  inclusion_rules: string[]
  exclusion_rules: string[]
  parent_code_id: string | null
  positive_example_annotation_ids: string[]
  negative_example_annotation_ids: string[]
}

export type TransitionCodebookEntryInput = {
  expected_version: number
  lifecycle: CodebookLifecycle
  related_code_ids: string[]
  reason: string
}

export type CreateAnalysisThemeInput = {
  label: string
  central_concept: string
  code_ids: string[]
  annotation_ids: string[]
}

export type CreateAnalysisMemoLinkInput = {
  memo_id: string
  target_kind: MemoTargetKind
  target_ref: string
  annotation_ids: string[]
}

export type SaveAnalysisCaseProfileInput = {
  expected_version: number | null
  case_ref: string
  display_label: string
  attributes: AnalysisCaseAttribute[]
  summary: string
  annotation_ids: string[]
  memo_ids: string[]
}

export type SaveCaseThemeMatrixCellInput = {
  expected_version: number | null
  case_profile_id: string
  subject_kind: MatrixSubjectKind
  subject_id: string
  summary: string
  annotation_ids: string[]
  memo_ids: string[]
  finding_kinds: ComparisonFindingKind[]
}

export type SetQualitativeMethodInput = {
  method: QualitativeMethod
  expected_version: number | null
}

File: backend/src/qunxue_api/adapters/sqlite/research_analysis_model.py
"""SQLite rows for task-scoped qualitative analysis records."""

from datetime import datetime

from sqlalchemy import (
    JSON,
    CheckConstraint,
    DateTime,
    ForeignKey,
    Index,
    Integer,
    String,
    Text,
    UniqueConstraint,
)
from sqlalchemy.orm import Mapped, mapped_column

from qunxue_api.adapters.sqlite.base import Base

_DECISION_STATUS_CHECK = "status IN ('candidate', 'confirmed', 'rejected')"
_DECISION_LIFECYCLE_CHECK = """
(
    status = 'candidate'
    AND version = 1
    AND decided_at IS NULL
    AND decision_reason IS NULL
)
OR
(
    status IN ('confirmed', 'rejected')
    AND version >= 2
    AND decided_at IS NOT NULL
    AND decision_reason IS NOT NULL
)
"""


class ResearchAnalysisWriteRequestRow(Base):
    __tablename__ = "research_analysis_write_requests"
    __table_args__ = (
        UniqueConstraint(
            "user_id",
            "task_id",
            "namespace",
            "idempotency_key",
            name="uq_research_analysis_write_requests_identity",
        ),
        CheckConstraint(
            "length(request_hash) = 64",
            name="ck_research_analysis_write_requests_hash",
        ),
    )

    request_id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    user_id: Mapped[str] = mapped_column(
        ForeignKey("users.user_id", ondelete="CASCADE"), nullable=False
    )
    task_id: Mapped[str] = mapped_column(
        ForeignKey("research_tasks.task_id", ondelete="CASCADE"), nullable=False
    )
    namespace: Mapped[str] = mapped_column(String(32), nullable=False)
    idempotency_key: Mapped[str] = mapped_column(String(512), nullable=False)
    operation: Mapped[str] = mapped_column(String(64), nullable=False)
    request_hash: Mapped[str] = mapped_column(String(64), nullable=False)
    result_kind: Mapped[str] = mapped_column(String(32), nullable=False)
    result_id: Mapped[str] = mapped_column(String(36), nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)


class ResearchAnnotationRow(Base):
    __tablename__ = "research_analysis_annotations"
    __table_args__ = (
        CheckConstraint(
            "annotation_kind IN ('descriptive', 'researcher_reflection')",
            name="ck_research_analysis_annotations_kind",
        ),
        CheckConstraint(
            "length(segment_content_hash) = 64",
            name="ck_research_analysis_annotations_segment_hash",
        ),
        CheckConstraint(
            "length(quote_hash) = 64",
            name="ck_research_analysis_annotations_quote_hash",
        ),
        CheckConstraint(
            "quote_start >= 0 AND quote_end > quote_start",
            name="ck_research_analysis_annotations_quote_range",
        ),
        CheckConstraint(
            "quote_end - quote_start = length(quote)",
            name="ck_research_analysis_annotations_quote_length",
        ),
        CheckConstraint(
            "annotation_kind != 'researcher_reflection' OR "
            "(reflection IS NOT NULL AND length(trim(reflection)) > 0)",
            name="ck_research_analysis_annotations_reflection_required",
        ),
        Index(
            "ix_research_analysis_annotations_owner_created",
            "user_id",
            "task_id",
            "created_at",
        ),
        Index(
            "ix_research_analysis_annotations_source",
            "material_id",
            "parse_id",
            "segment_id",
        ),
    )

    annotation_id: Mapped[str] = mapped_column(String(36), primary_key=True)
    user_id: Mapped[str] = mapped_column(
        ForeignKey("users.user_id", ondelete="CASCADE"), nullable=False
    )
    task_id: Mapped[str] = mapped_column(
        ForeignKey("research_tasks.task_id", ondelete="CASCADE"), nullable=False
    )
    material_id: Mapped[str] = mapped_column(
        ForeignKey("research_materials.material_id", ondelete="RESTRICT"), nullable=False
    )
    parse_id: Mapped[str] = mapped_column(String(36), nullable=False)
    segment_id: Mapped[str] = mapped_column(String(128), nullable=False)
    segment_content_hash: Mapped[str] = mapped_column(String(64), nullable=False)
    quote: Mapped[str] = mapped_column(Text, nullable=False)
    quote_hash: Mapped[str] = mapped_column(String(64), nullable=False)
    quote_start: Mapped[int] = mapped_column(Integer, nullable=False)
    quote_end: Mapped[int] = mapped_column(Integer, nullable=False)
    locator: Mapped[dict[str, object]] = mapped_column(JSON, nullable=False)
    annotation_kind: Mapped[str] = mapped_column(String(32), nullable=False)
    case_label: Mapped[str | None] = mapped_column(String(256), nullable=True)
    observed_at: Mapped[str | None] = mapped_column(String(128), nullable=True)
    note: Mapped[str] = mapped_column(Text, nullable=False)
    reflection: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)


class ResearchCodeRow(Base):
    __tablename__ = "research_analysis_codes"
    __table_args__ = (
        CheckConstraint(
            _DECISION_STATUS_CHECK,
            name="ck_research_analysis_codes_status",
        ),
        CheckConstraint(
            _DECISION_LIFECYCLE_CHECK,
            name="ck_research_analysis_codes_lifecycle",
        ),
        Index(
            "ix_research_analysis_codes_owner_created",
            "user_id",
            "task_id",
            "created_at",
        ),
    )

    code_id: Mapped[str] = mapped_column(String(36), primary_key=True)
    user_id: Mapped[str] = mapped_column(
        ForeignKey("users.user_id", ondelete="CASCADE"), nullable=False
    )
    task_id: Mapped[str] = mapped_column(
        ForeignKey("research_tasks.task_id", ondelete="CASCADE"), nullable=False
    )
    label: Mapped[str] = mapped_column(String(256), nullable=False)
    definition: Mapped[str] = mapped_column(Text, nullable=False)
    annotation_ids: Mapped[list[str]] = mapped_column(JSON, nullable=False)
    rationale: Mapped[str] = mapped_column(Text, nullable=False)
    source: Mapped[str] = mapped_column(String(32), nullable=False)
    status: Mapped[str] = mapped_column(String(32), nullable=False)
    version: Mapped[int] = mapped_column(Integer, nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    conversation_id: Mapped[str | None] = mapped_column(String(36), nullable=True)
    agent_run_id: Mapped[str | None] = mapped_column(String(36), nullable=True)
    agent_turn_id: Mapped[str | None] = mapped_column(String(36), nullable=True)
    tool_call_id: Mapped[str | None] = mapped_column(String(512), nullable=True)
    decided_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    decision_reason: Mapped[str | None] = mapped_column(Text, nullable=True)


class ResearchMemoRow(Base):
    __tablename__ = "research_analysis_memos"
    __table_args__ = (
        CheckConstraint(
            "memo_kind IN ('descriptive', 'reflexive', 'analytic', 'methodological')",
            name="ck_research_analysis_memos_kind",
        ),
        CheckConstraint(
            _DECISION_STATUS_CHECK,
            name="ck_research_analysis_memos_status",
        ),
        CheckConstraint(
            _DECISION_LIFECYCLE_CHECK,
            name="ck_research_analysis_memos_lifecycle",
        ),
        Index(
            "ix_research_analysis_memos_owner_created",
            "user_id",
            "task_id",
            "created_at",
        ),
    )

    memo_id: Mapped[str] = mapped_column(String(36), primary_key=True)
    user_id: Mapped[str] = mapped_column(
        ForeignKey("users.user_id", ondelete="CASCADE"), nullable=False
    )
    task_id: Mapped[str] = mapped_column(
        ForeignKey("research_tasks.task_id", ondelete="CASCADE"), nullable=False
    )
    title: Mapped[str] = mapped_column(String(512), nullable=False)
    content: Mapped[str] = mapped_column(Text, nullable=False)
    memo_kind: Mapped[str] = mapped_column(String(32), nullable=False)
    annotation_ids: Mapped[list[str]] = mapped_column(JSON, nullable=False)
    code_ids: Mapped[list[str]] = mapped_column(JSON, nullable=False)
    source: Mapped[str] = mapped_column(String(32), nullable=False)
    status: Mapped[str] = mapped_column(String(32), nullable=False)
    version: Mapped[int] = mapped_column(Integer, nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    conversation_id: Mapped[str | None] = mapped_column(String(36), nullable=True)
    agent_run_id: Mapped[str | None] = mapped_column(String(36), nullable=True)
    agent_turn_id: Mapped[str | None] = mapped_column(String(36), nullable=True)
    tool_call_id: Mapped[str | None] = mapped_column(String(512), nullable=True)
    decided_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    decision_reason: Mapped[str | None] = mapped_column(Text, nullable=True)


class ResearchComparisonRow(Base):
    __tablename__ = "research_analysis_comparisons"
    __table_args__ = (
        CheckConstraint(
            _DECISION_STATUS_CHECK,
            name="ck_research_analysis_comparisons_status",
        ),
        CheckConstraint(
            _DECISION_LIFECYCLE_CHECK,
            name="ck_research_analysis_comparisons_lifecycle",
        ),
        Index(
            "ix_research_analysis_comparisons_owner_created",
            "user_id",
            "task_id",
            "created_at",
        ),
    )

    comparison_id: Mapped[str] = mapped_column(String(36), primary_key=True)
    user_id: Mapped[str] = mapped_column(
        ForeignKey("users.user_id", ondelete="CASCADE"), nullable=False
    )
    task_id: Mapped[str] = mapped_column(
        ForeignKey("research_tasks.task_id", ondelete="CASCADE"), nullable=False
    )
    title: Mapped[str] = mapped_column(String(512), nullable=False)
    question: Mapped[str] = mapped_column(Text, nullable=False)
    case_labels: Mapped[list[str]] = mapped_column(JSON, nullable=False)
    time_labels: Mapped[list[str]] = mapped_column(JSON, nullable=False)
    findings: Mapped[list[dict[str, object]]] = mapped_column(JSON, nullable=False)
    competing_explanations: Mapped[list[str]] = mapped_column(JSON, nullable=False)
    evidence_gaps: Mapped[list[str]] = mapped_column(JSON, nullable=False)
    next_steps: Mapped[list[dict[str, object]]] = mapped_column(JSON, nullable=False)
    theory_implication: Mapped[str] = mapped_column(Text, nullable=False)
    source: Mapped[str] = mapped_column(String(32), nullable=False)
    status: Mapped[str] = mapped_column(String(32), nullable=False)
    version: Mapped[int] = mapped_column(Integer, nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    conversation_id: Mapped[str | None] = mapped_column(String(36), nullable=True)
    agent_run_id: Mapped[str | None] = mapped_column(String(36), nullable=True)
    agent_turn_id: Mapped[str | None] = mapped_column(String(36), nullable=True)
    tool_call_id: Mapped[str | None] = mapped_column(String(512), nullable=True)
    decided_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    decision_reason: Mapped[str | None] = mapped_column(Text, nullable=True)


class ResearchCodebookEntryRow(Base):
    __tablename__ = "research_analysis_codebook_entries"
    __table_args__ = (
        CheckConstraint(
            "lifecycle IN ('active', 'merged', 'split', 'retired')",
            name="ck_research_analysis_codebook_lifecycle",
        ),
        Index(
            "ix_research_analysis_codebook_owner",
            "user_id",
            "task_id",
            "updated_at",
        ),
    )

    code_id: Mapped[str] = mapped_column(
        ForeignKey("research_analysis_codes.code_id", ondelete="CASCADE"),
        primary_key=True,
    )
    user_id: Mapped[str] = mapped_column(
        ForeignKey("users.user_id", ondelete="CASCADE"), nullable=False
    )
    task_id: Mapped[str] = mapped_column(
        ForeignKey("research_tasks.task_id", ondelete="CASCADE"), nullable=False
    )
    inclusion_rules: Mapped[list[str]] = mapped_column(JSON, nullable=False)
    exclusion_rules: Mapped[list[str]] = mapped_column(JSON, nullable=False)
    parent_code_id: Mapped[str | None] = mapped_column(String(36), nullable=True)
    positive_example_annotation_ids: Mapped[list[str]] = mapped_column(JSON, nullable=False)
    negative_example_annotation_ids: Mapped[list[str]] = mapped_column(JSON, nullable=False)
    lifecycle: Mapped[str] = mapped_column(String(32), nullable=False)
    related_code_ids: Mapped[list[str]] = mapped_column(JSON, nullable=False)
    version: Mapped[int] = mapped_column(Integer, nullable=False)
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    revision_reason: Mapped[str] = mapped_column(Text, nullable=False)


class ResearchThemeRow(Base):
    __tablename__ = "research_analysis_themes"
    __table_args__ = (
        CheckConstraint(_DECISION_STATUS_CHECK, name="ck_research_analysis_themes_status"),
        CheckConstraint(_DECISION_LIFECYCLE_CHECK, name="ck_research_analysis_themes_lifecycle"),
        Index("ix_research_analysis_themes_owner", "user_id", "task_id", "created_at"),
    )

    theme_id: Mapped[str] = mapped_column(String(36), primary_key=True)
    user_id: Mapped[str] = mapped_column(
        ForeignKey("users.user_id", ondelete="CASCADE"), nullable=False
    )
    task_id: Mapped[str] = mapped_column(
        ForeignKey("research_tasks.task_id", ondelete="CASCADE"), nullable=False
    )
    label: Mapped[str] = mapped_column(String(512), nullable=False)
    central_concept: Mapped[str] = mapped_column(Text, nullable=False)
    code_ids: Mapped[list[str]] = mapped_column(JSON, nullable=False)
    annotation_ids: Mapped[list[str]] = mapped_column(JSON, nullable=False)
    source: Mapped[str] = mapped_column(String(32), nullable=False)
    status: Mapped[str] = mapped_column(String(32), nullable=False)
    version: Mapped[int] = mapped_column(Integer, nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    decided_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    decision_reason: Mapped[str | None] = mapped_column(Text, nullable=True)


class ResearchMemoLinkRow(Base):
    __tablename__ = "research_analysis_memo_links"
    __table_args__ = (
        CheckConstraint(
            "target_kind IN "
            "('project', 'material', 'source', 'code', 'case', 'comparison', 'draft')",
            name="ck_research_analysis_memo_links_target_kind",
        ),
        Index("ix_research_analysis_memo_links_owner", "user_id", "task_id", "created_at"),
    )

    link_id: Mapped[str] = mapped_column(String(36), primary_key=True)
    user_id: Mapped[str] = mapped_column(
        ForeignKey("users.user_id", ondelete="CASCADE"), nullable=False
    )
    task_id: Mapped[str] = mapped_column(
        ForeignKey("research_tasks.task_id", ondelete="CASCADE"), nullable=False
    )
    memo_id: Mapped[str] = mapped_column(
        ForeignKey("research_analysis_memos.memo_id", ondelete="CASCADE"), nullable=False
    )
    target_kind: Mapped[str] = mapped_column(String(32), nullable=False)
    target_ref: Mapped[str] = mapped_column(String(512), nullable=False)
    annotation_ids: Mapped[list[str]] = mapped_column(JSON, nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)


class ResearchCaseProfileRow(Base):
    __tablename__ = "research_analysis_case_profiles"
    __table_args__ = (
        UniqueConstraint(
            "user_id",
            "task_id",
            "case_ref",
            name="uq_research_analysis_case_profiles_ref",
        ),
        Index("ix_research_analysis_case_profiles_owner", "user_id", "task_id", "updated_at"),
    )

    profile_id: Mapped[str] = mapped_column(String(36), primary_key=True)
    user_id: Mapped[str] = mapped_column(
        ForeignKey("users.user_id", ondelete="CASCADE"), nullable=False
    )
    task_id: Mapped[str] = mapped_column(
        ForeignKey("research_tasks.task_id", ondelete="CASCADE"), nullable=False
    )
    case_ref: Mapped[str] = mapped_column(String(512), nullable=False)
    display_label: Mapped[str] = mapped_column(String(512), nullable=False)
    attributes: Mapped[list[list[str]]] = mapped_column(JSON, nullable=False)
    summary: Mapped[str] = mapped_column(Text, nullable=False)
    annotation_ids: Mapped[list[str]] = mapped_column(JSON, nullable=False)
    memo_ids: Mapped[list[str]] = mapped_column(JSON, nullable=False)
    version: Mapped[int] = mapped_column(Integer, nullable=False)
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)


class ResearchMatrixCellRow(Base):
    __tablename__ = "research_analysis_matrix_cells"
    __table_args__ = (
        UniqueConstraint(
            "user_id",
            "task_id",
            "case_profile_id",
            "subject_kind",
            "subject_id",
            name="uq_research_analysis_matrix_cells_position",
        ),
        CheckConstraint(
            "subject_kind IN ('code', 'theme')",
            name="ck_research_analysis_matrix_cells_subject_kind",
        ),
        Index("ix_research_analysis_matrix_cells_owner", "user_id", "task_id", "updated_at"),
    )

    cell_id: Mapped[str] = mapped_column(String(36), primary_key=True)
    user_id: Mapped[str] = mapped_column(
        ForeignKey("users.user_id", ondelete="CASCADE"), nullable=False
    )
    task_id: Mapped[str] = mapped_column(
        ForeignKey("research_tasks.task_id", ondelete="CASCADE"), nullable=False
    )
    case_profile_id: Mapped[str] = mapped_column(
        ForeignKey("research_analysis_case_profiles.profile_id", ondelete="CASCADE"),
        nullable=False,
    )
    subject_kind: Mapped[str] = mapped_column(String(32), nullable=False)
    subject_id: Mapped[str] = mapped_column(String(36), nullable=False)
    summary: Mapped[str] = mapped_column(Text, nullable=False)
    annotation_ids: Mapped[list[str]] = mapped_column(JSON, nullable=False)
    memo_ids: Mapped[list[str]] = mapped_column(JSON, nullable=False)
    finding_kinds: Mapped[list[str]] = mapped_column(JSON, nullable=False)
    version: Mapped[int] = mapped_column(Integer, nullable=False)
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)


class ResearchMethodPresetRow(Base):
    __tablename__ = "research_analysis_method_presets"

    user_id: Mapped[str] = mapped_column(
        ForeignKey("users.user_id", ondelete="CASCADE"), primary_key=True
    )
    task_id: Mapped[str] = mapped_column(
        ForeignKey("research_tasks.task_id", ondelete="CASCADE"), primary_key=True
    )
    method: Mapped[str] = mapped_column(String(64), nullable=False)
    version: Mapped[int] = mapped_column(Integer, nullable=False)
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
